main/lsp/requests/rename.cc

```diff
@@ -29,47 +29,208 @@ bool isValidRenameLocation(const core::SymbolRef &symbol, const core::GlobalStat
+// Follow superClass links until we find the highest class that contains the given method. In other words we find the
+// "root" of the tree of classes that define a method.
+core::SymbolRef findRootClassWithMethod(const core::GlobalState &gs, core::SymbolRef klass, core::NameRef methodName) {
+    auto root = klass;
+    ENFORCE(klass.data(gs)->isClassOrModule());
+    while (true) {
+        auto tmp = root.data(gs)->superClass();
+        ENFORCE(tmp.exists()); // everything derives from Kernel::Object so we can't ever reach the actual top type
+        if (!tmp.exists() || !(tmp.data(gs)->findMember(gs, methodName).exists())) {
+            break;
+        }
+        root = tmp;
     }
+    return root;
+}
+
+class Renamer {
+public:
+    Renamer(const string oldName, const string newName) : oldName(oldName), newName(newName) {}
+    virtual ~Renamer() = default;
+    virtual void rename(const core::GlobalState &gs, unique_ptr<Location> &location, core::FileRef fref) = 0;
+    unique_ptr<WorkspaceEdit> buildEdit();
 
+protected:
+    string oldName;
+    string newName;
+    UnorderedMap<string, vector<unique_ptr<TextEdit>>> edits;
+};
+
+unique_ptr<WorkspaceEdit> Renamer::buildEdit() {
+    auto we = make_unique<WorkspaceEdit>();
     vector<unique_ptr<TextDocumentEdit>> textDocEdits;
     for (auto &item : edits) {
         // TODO: Version.
         textDocEdits.push_back(make_unique<TextDocumentEdit>(
             make_unique<VersionedTextDocumentIdentifier>(item.first, JSONNullObject()), move(item.second)));
     }
     we->documentChanges = move(textDocEdits);
-
     return we;
 }
+class MethodRenamer : public Renamer {
+public:
+    MethodRenamer(const string oldName, const string newName) : Renamer(oldName, newName) {}
+    ~MethodRenamer() {}
+    void rename(const core::GlobalState &gs, unique_ptr<Location> &location, core::FileRef fref) override {
+        auto loc = location->range->toLoc(gs, fref);
+        auto source = loc.source(gs);
+        string newsrc;
+        if (absl::StartsWith(source, "def ") || absl::StartsWith(source, "attr_reader") ||
+            absl::StartsWith(source, "attr_accessor")) {
+            newsrc = replaceMethodNameInDef(source);
+        } else { // send
+            newsrc = replaceMethodNameInSend(source);
+        }
+        edits[location->uri].push_back(make_unique<TextEdit>(move(location->range), newsrc));
+    }
+
+private:
+    // Parse a method name from a send (call), replace it with the new name, and return the full send source with the
+    // new name
+    string replaceMethodNameInSend(string send) {
+        // Split into method and args
+        string methodExpr, args;
+        for (string::size_type i = 0; i < send.size(); i++) {
+            if (send[i] == ' ' || send[i] == '(') {
+                methodExpr = send.substr(0, i);
+                args = send.substr(i, send.size() - i);
```

Comment:
Should this break out of the loop?

Also, does this work on callsites with a receiver? e.g.:

```ruby
  foo.bar(a, b, c)
#  ^^^ this is not the method
#.    ^^^ this is the method
```

is `foo` part of the send loc? It _looks like_ the tests pass, so maybe this isn't a problem.

Reply:
Nit: Actually, maybe we could replace this with something simpler, like `send.find(methodName)` that verifies that the next token in `send` (if present) isn't alphanumeric? Up to you if you want to do it that way.

Reply:
Yes it should break from the loop, will fix that. I wanted to avoid searching for methodName because then we have to handle cases where it appears in the string but isn't actually the method we're looking for (like in the method_names test case)

Reply:
oh yeah! d'oh. makes sense :)

Reply:
And yes, `foo` is part of the send loc for `foo.bar(a,b,c)`. But we only break from the loop on `(` or space, so we get `foo.bar` in methodExpr

Reply:
After discussing in DM about other corner cases, we will fix this issue in a subsequent PR by surfacing receiver loc information in `SendQueryResponse`s. That will make it easy to exclude the part of the send loc that corresponds to the receiving object.